lib/wrap-request.js

```diff
@@ -0,0 +1,59 @@
+/* eslint padded-blocks: 0 */
+module.exports = wrapRequest
+
+const noop = () => Promise.resolve()
+
+function wrapRequest (state, emitter, request, options, retries = 0) {
+  const retryRequest = function (retryAfter) {
+    return new Promise(resolve => setTimeout(resolve, retryAfter * 1000))
+      .then(() => wrapRequest(state, emitter, request, options, retries + 1))
+  }
+
+  const jobOptions = retries > 0 ? { weight: 0, priority: 0 } : {}
+  return state.globalLimiter.schedule(jobOptions, async function () {
+
+    // Guarantee at least 1000ms between writes
+    if (options.method !== 'GET' && options.method !== 'HEAD') {
+      await state.writeLimiter.schedule(noop)
+    }
+
+    // Guarantee at least 5000ms between requests that trigger notifications
+    if (state.triggersNotification[options.url]) {
+      await state.triggersNotificationLimiter.schedule(noop)
+    }
+
+    try {
+      // Execute request
+      return await request(options)
+    } catch (error) {
+      if (error.status === 403 && error.message.toLowerCase().includes('abuse')) {
+        // The user has hit the abuse rate limit.
+        // https://developer.github.com/v3/#abuse-rate-limits
+
+        // Due to known issues with replication lag, the Retry-After header can sometimes be blank
+        // for a few seconds after entering 'abuse limited mode'.
+        // We ensure the Retry-After is always at least 5 seconds.
```

Comment:
You mention that the header is always present after 2-3 seconds. Maybe mention that in the comment instead of "for a few seconds"? Maybe something like

```js
        // The Retry-After header can sometimes be blank when hitting an abuse limit,
        // but is always present after 2-3s, so make sure to set `retryAfter` to at least 5s by default.
```

I’m not sure we should mention "known issues with replication lag" unless we can reference a resource with more information about it

Reply:
Good points! I'm changing it.